pkg/cache/cluster.go

```diff
@@ -1113,26 +1113,51 @@ func (c *clusterCache) processEvent(event watch.EventType, un *unstructured.Unst
 		return
 	}
 
+	ok, newRes, oldRes, ns := c.updateResource(key, event, un)
+	// Still requesting read lock since namespace resources can have a race of being used and being updated.
+	// Some race is still possible between releasing the lock and acquiring this, but since each resource type
+	// is processed by one goroutine, at most we'd get a fresher namespace resources, which should be fine,
+	// at least for existing resource update handler.
```

Comment:
I think we should either decide, **yes**, we can call the onResourceUpdated handlers with stale ns data and the handlers must handle it, or **no**, we need to retain the lock. We cannot really say that it "should be fine". It is very likely that other go routines will obtain the lock when it is released and before it is acquired again. There are no ways to predict as far as I know what will be the next resources events being processed and they could very well be on the same resource.

For instance, a finalizer is completed and removed from the finalizer array (update event), then the resource is deleted (delete event). What happens if the updated handler is called and the resource is not in the namespace map anymore because it just got deleted?

I am pretty sure it will be ok, but it needs to be 100% sure and the comment should reflect that `ns` is the last known namespace cache, and not the cache at the time of the event.

Reply:
> they could very well be on the same resource.
Watches are per object group/kind IIRC, so updates to the same resource would be handled by the same goroutine, so at least we don't have that part of a problem.

I agree that comment has to be better. We can definitely have OnResourceUpdated operating with a fresher view of namespace resources. Namespace resources are used to derive the application, so if that's updated, we'd get a new event anyways.

Reply:
I've also tested on live environment for a few days and don't see related issues so far. I'm quite confident that's okay.

Reply:
I'm not 100% what happens when the cluster gets refreshed, but also think it'll be okay since whole namespace resources slice should be still held by the goroutine.